File: examples/python/testing-patterns/complex-workflows/maximum-attempts-workflow.py
@durable_execution
def handler(event: dict, context: DurableContext) -> dict:
    target = event.get("target", 10)
    state = 0
    attempt = 0
    max_attempts = 5
    
    while attempt < max_attempts and state < target:
        attempt += 1
        state = context.step(lambda _, s=state: s + 1, name=f"attempt_{attempt}")
        
        if state < target:
            context.wait(Duration.from_seconds(1), name=f"wait_{attempt}")
    
    return {"state": state, "attempts": attempt, "reached_target": state >= target}

File: examples/java/testing-patterns/complex-workflows/maximum-attempts-workflow.java
// Maximum attempts workflow handler
public class MaxAttemptsHandler extends DurableHandler<Map, Map> {
    @Override
    public Map handleRequest(Map input, DurableContext ctx) {
        int target = (int) input.getOrDefault("target", 10);
        int[] state = {0};
        int[] attempt = {0};
        int maxAttempts = 5;

        while (attempt[0] < maxAttempts && state[0] < target) {
            attempt[0]++;
            int currentState = state[0];
            int currentAttempt = attempt[0];
            state[0] = ctx.step("attempt_" + currentAttempt, Integer.class,
                stepCtx -> currentState + 1);

            if (state[0] < target) {
                ctx.wait("wait_" + currentAttempt, Duration.ofSeconds(1));
            }
        }

        return Map.of(
            "state", state[0],
            "attempts", attempt[0],
            "reached_target", state[0] >= target
        );
    }
}
